microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/validators/annotations/RangeExpression.java

```diff
@@ -0,0 +1,220 @@
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.jdt.internal.core.java.validators.annotations;
+
+import java.text.MessageFormat;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * The range expression used to validate value from an attribute annotation.
+ * 
+ * The range is specified like OSGi version syntax:
+ * 
+ * <ul>
+ * <li>0 -> means >=0</li>
+ * <li>[0 -> means >=0</li>
+ * <li>(0 -> means >0</li>
+ * <li>[0,1] -> means >=0 && <=1</li>
+ * <li>(0,1] -> means >0 && <=1</li>
+ * <li>(0,1) -> means >0 && <1</li>
+ * <li>[0,1) -> means >=0 && <1</li>
+ * </ul>
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class RangeExpression {
+
+	private final Edge from;
+	private final Edge to;
+	private final String errorMessage;
+
+	public RangeExpression(Edge from, Edge to) {
+		this.from = from;
+		this.to = to;
+		this.errorMessage = createErrorMessage(from, to);
+	}
+
+	private static String createErrorMessage(Edge from, Edge to) {
+		StringBuilder message = new StringBuilder("The value ''{0}'' must be ");
+		if (to == null) {
+			message.append("greater than ");
+			if (from.inclusive) {
+				message.append("or equal to ");
+			}
+			message.append("''");
+			message.append(from.valueAsString);
+			message.append("''.");
+		} else {
+			// The value must be between 0 and 1 inclusive.
+			message.append("between ");
+			message.append("''");
+			message.append(from.valueAsString);
+			message.append("''");
+			if (from.inclusive) {
+				message.append(" inclusive");
+			} else {
+				message.append(" exclusive");
+			}
+			message.append(" and ");
+			message.append("''");
+			message.append(to.valueAsString);
+			message.append("''");
+			if (to.inclusive) {
+				message.append(" inclusive");
+			} else {
+				message.append(" exclusive");
+			}
+			message.append(".");
+		}
+		return message.toString();
+	}
+
+	public String validate(double value) {
+		if (!from.validate(value) || (to != null && !to.validate(value))) {
+			return MessageFormat.format(errorMessage, value);
+		}
+		return null;
+	}
+
+	/**
+	 * Parse the given range <code>expression</code>.
+	 * 
+	 * @param expression the range expression to parse.
+	 * @return an instance of range expression.
+	 * @throws RangeExpressionException thrown when expression doesn't respect the
+	 *                                  range expression syntax.
+	 */
+	public static RangeExpression parse(String expression) throws RangeExpressionException {
+		if (StringUtils.isEmpty(expression)) {
+			throw new RangeExpressionException("No expression");
+		}
+		Edge from = null;
+		Edge to = null;
+
+		boolean inclusive = true;
+		StringBuilder valueAsString = new StringBuilder();
+
+		for (int i = 0; i < expression.length(); i++) {
+			char c = expression.charAt(i);
+			switch (c) {
+			case ' ':
+				// ignore whitespace
+				break;
+			case '[':

```

Comment:
Is it worth also supporting `0)` or `0]` if we do the opposite ?

Reply:
The both if I understand your comments. See Javadoc https://github.com/eclipse/lsp4mp/pull/173/files#diff-3e4d6d0d761c9ffc6f9f9d20379ec87c91773ab5778f245110280ad427691a5fR26

Reply:
I meant, just as we support `(0` to mean `> 0`, is it worth supporting `0)` to mean `<0` ? I changed one of the tests to try this syntax but it would seem to fail. It's pretty minor as I can't think of many circumstances where this would be needed.

Reply:
Ok I understand, if we have this usecase, I will work on it.